src/WebJobs.Script.WebHost/Diagnostics/WebHostMetricsLogger.cs

```diff
@@ -20,33 +21,71 @@ public WebHostMetricsLogger(IEventGenerator eventGenerator, int metricEventInter
             _metricsEventManager = new MetricsEventManager(eventGenerator, metricEventIntervalInSeconds);
         }
 
+        public object BeginEvent(string eventName)
+        {
+            return _metricsEventManager.BeginEvent(eventName);
+        }
+
         public void BeginEvent(MetricEvent metricEvent)
         {
-            FunctionStartedEvent startedEvent = metricEvent as FunctionStartedEvent;
-            if (startedEvent != null)
+            if (metricEvent is FunctionStartedEvent)
             {
-                startedEvent.StartTime = DateTime.Now;
+                FunctionStartedEvent startedEvent = metricEvent as FunctionStartedEvent;
+                startedEvent.Timestamp = DateTime.Now;
                 _metricsEventManager.FunctionStarted(startedEvent);
             }
         }
 
+        public void EndEvent(object eventHandle)
+        {
+            _metricsEventManager.EndEvent(eventHandle);
+        }
+
         public void EndEvent(MetricEvent metricEvent)
         {
-            FunctionStartedEvent completedEvent = metricEvent as FunctionStartedEvent;
-            if (completedEvent != null)
+            if (metricEvent is FunctionStartedEvent)
             {
-                completedEvent.EndTime = DateTime.Now;
+                FunctionStartedEvent completedEvent = metricEvent as FunctionStartedEvent;
+                completedEvent.Duration = DateTime.Now - completedEvent.Timestamp;
                 _metricsEventManager.FunctionCompleted(completedEvent);
             }
         }
 
         public void LogEvent(MetricEvent metricEvent)
         {
-            HostStarted hostStartedEvent = metricEvent as HostStarted;
-            if (hostStartedEvent != null)
+            if (metricEvent is HostStarted)
```

Comment:
Should be able to avoid the double cast here by doing an `as HostStarted` and doing a check